Fix build loop

Profpatsch opened the pull request:
fix(build_loop): always schedule at most one other build
80e8eb1

This switches the build loop from blockingly running a build for every
event that ever appeared, to always at most running one additional
build as soon as the current one finishes.

Whenver we start a new build, we take the *current* state of the
environment to run nix, which means that all events that happen
while a build is running (e.g. pings, filesystem events &c.) should be
collapsed into *one* (1) extra build to be scheduled right after the
current build finishes.

So there is now a stateful `current_build` variable, which contains
one `BuildState` that is mutably updated from `NotRunning` to
`Running` and if it’s already running to `RunningAndScheduled`. Then
the “build finished” event becomes just another handler in our giant
`select!`, meaning that as soon as a build finishes we can either go
back to `NotRunning` or go to `Running` if there had been another
build scheduled.

For now this means inlining `once_with_send` into the build loop,
further refactors to follow to make the logic easier to understand.



### Checklist

- [ ] Updated the documentation (code documentation, command help, ...)
- [ ] Tested the change (unit or integration tests)
- [ ] Amended the changelog in `release.nix` (see `release.nix` for instructions)

<!-- This checklist is here to help you and your reviewers, so feel free to edit it as appropriate,
e.g. bugfixes don't usually require a documentation change. -->

Profpatsch committed:
feat(run_async): add Async

Small helper struct that abstracts over the common way of running a
function asynchronously in a helper thread. Can either blockingly wait
on the result or use a channel to read it.
The API is not entirely watertight, but it doesn’t really matter.

Profpatsch committed:
chore(build_loop): remove mystical output gc dir check

This check for whether the gc output path exists has been there in
forever, undocumented. Then I added a TODO asking for clarification
another eternity ago, but even staring at it for extended amounts of
time does not give me any clues as to why it is there.

I say in face of the obvious simplification of the code we can remove
it for now, the only effect it could have is rebuilding more often.
But since we are *already* rebuilding way too often (until we improve
the build loop which I’m in the process of doing), we can risk that.
Simple code is more important for the restructure of the build loop.

Profpatsch committed:
fix(build_loop): always schedule at most one other build

This switches the build loop from blockingly running a build for every
event that ever appeared, to always at most running one additional
build as soon as the current one finishes.

Whenver we start a new build, we take the *current* state of the
environment to run nix, which means that all events that happen
while a build is running (e.g. pings, filesystem events &c.) should be
collapsed into *one* (1) extra build to be scheduled right after the
current build finishes.

So there is now a stateful `current_build` variable, which contains
one `BuildState` that is mutably updated from `NotRunning` to
`Running` and if it’s already running to `RunningAndScheduled`. Then
the “build finished” event becomes just another handler in our giant
`select!`, meaning that as soon as a build finishes we can either go
back to `NotRunning` or go to `Running` if there had been another
build scheduled.

For now this means inlining `once_with_send` into the build loop,
further refactors to follow to make the logic easier to understand.

Profpatsch committed:
fix(build_loop): remove initial unconditional build

The build loop would previously drain pings in order to prevent
building the project twice immediately. That is well-meaning, but just
a workaround for the fact, that we would already build once.

However, we can remove the unconditional first build, since external
processes would ping us anyway. That is, the build loop would not even
be started unless some external process has pinged the daemon.

This leads to `lorri watch` not building initially, but we can just do
the same as external clients would do, and ping the build loop.

Profpatsch committed:
chore(run_async): bump test timeout

The recv should be immediate, but if the system is under load the
threads might be scheduled differently and 1 ms might have passed,
leading to the recv timing out.

Profpatsch commented:
@symphorien, can you try this out, I think this supersedes https://github.com/nix-community/lorri/pull/3 in a more principled way, while also simplifying the whole thing considerably.

Profpatsch committed:
refact(build_loop): directly schedule builds in select!()

Returning a reason and scheduling at the very end was an artifact from
before consolidation into one `select!()`.

The flow is a lot clearer if we just directly schedule.

sternenseemann commented on src/run_async.rs:
```diff
@@ -0,0 +1,117 @@
+            Err(chan::RecvError) =>
+                panic!("unable to receive the async result, because the channel was disconnected and empty (should never happen)")
+        };
+        // since the function finished, the thread (created in `run()`) will join immediatly after
```
since the function _is_ finished

Profpatsch replied:
Nah, it finished (active)

symphorien commented:
~I can still reproduce https://github.com/target/lorri/issues/464 on top of this PR.~

symphorien commented:
Ah no sorry, I had my timing off, it works.

Profpatsch committed:
fix(ops/upgrade): display changelog in reverse

before it would display the entries in reverse order, so the newest
events would appear first, and scroll off the screen if there’s too
many.

Profpatsch committed:
doc(release.nix): add changelog for the build loop fix

Profpatsch commented:
Okay, added the change log, will merge and then we can release a new version upstream.

symphorien commented:
It looks like this builds the project on every ping.

sternenseemann commented:
> It looks like this builds the project on every ping.

I didn't encounter that behavior.

symphorien commented:
See https://github.com/nix-community/lorri/issues/11 for reproducer.
